src/nodes.coffee

```diff
@@ -4166,28 +4186,24 @@ makeDelimitedLiteral = (body, options = {}) ->
     when other     then (if options.double then "\\#{other}" else other)
   "#{options.delimiter}#{body}#{options.delimiter}"
 
-# Extends the location data of an AST node to include the location data from
-# another AST node.
-mergeAstLocationData = (intoNode, fromNode) ->
-  {range: intoRange} = intoNode
-  {range: fromRange} = fromNode
-  return intoNode unless intoRange and fromRange
-  if fromRange[0] < intoRange[0]
-    intoNode.range = intoRange = [
-      fromRange[0]
-      intoRange[1]
-    ]
-    intoNode.start = fromNode.start
-    intoNode.loc =
-      start: fromNode.loc.start
-      end: intoNode.loc.end
-  if fromRange[1] > intoRange[1]
-    intoNode.range = [
-      intoRange[0]
-      fromRange[1]
+# Take two AST nodes, or two AST nodes’ location data objects, and return a new
+# location data object that encompasses the location data of both nodes. So the
+# new `start` value will be the earlier of the two nodes’ `start` values, the
+# new `end` value will be the later of the two nodes’ `end` values, etc.
+lesser  = (a, b) -> if a < b then a else b
+greater = (a, b) -> if a > b then a else b
+mergeAstLocationData = (nodeA, nodeB) ->
```

Comment:
I can see separating this "calculation of the new location data" out like this, but I think it'll still be clunky not to have a wrapper function that mutates an existing AST node. But I'd have to see it in use eg in an overridden `ast()` method on `Value`

I question your aversion to "mutating an AST object that gets passed around". To an extent I can see how it's a little more obvious what's going on if you're always visibly assembling pieces of the AST eg in the `ast()` method. But your claim that it'd be "hell to debug" or whatever is simply not true in my experience, I've been doing heavy development both of and against this AST code for months and there have been about zero times that I got confused about where a specific AST property was getting set. Having a few "known places" like `mergeAstLocationData()` that mutate AST objects in a predictable way does not seem like a violation of basic principles that will result in unmaintainable code